### PR TITLE
[Websocket] Improve Topic handling

### LINKED ISSUE #2750: Websocket Topics does not consider URI Mapping
### Description:
Instead of considering URI Mapping, the topic of the API is always appended to the Backend URL. Via fixing this,

- Users have control of the topic needs to be appended to backend (path rewrite)
- Users can transform the websocket path parameters into query params using {var.uri.<path-parameter>} syntax.

### Steps to reproduce:
Create a websocket API and add a topic with a different URI Mapping.
Try invoking the API resource.

### Affected Product Version:
1.1.0-alpha

### Environment details (with versions):
- OS:
- Client:
- Env (Docker/K8s):

---

### LINKED ISSUE #2780: [Websockets] Memory Leak due to accumulating stream_handlers.
### Description:
Due to $subject, the Wasm VM could exit at some point, when using websocket APIs. Following error is visible when this happens.

`[2022-03-21 18:01:52.200][16][error][wasm] [source/extensions/common/wasm/wasm_vm.cc:38] Function: proxy_on_grpc_receive failed: Uncaught RuntimeError: memory access out of bounds`

### Affected Product Version:
1.0.0, 1.1.0-alpha

### Environment details (with versions):
- OS:
- Client:
- Env (Docker/K8s):

---

### PR DESCRIPTION
**Pull request by VirajSalaka**
### Purpose

Users can do path rewrite, path parameter to query parameter conversion when supporting multiple topics for Websocket APIs.

### Issues

Fixes #2750 
Fixes #2780

### Automation tests
 - Unit tests added: No
 - Integration tests added: Yes

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java [hunk 1]
```diff
@@ -85,6 +91,7 @@ private boolean applyPolicy(RequestContext requestContext, Policy policy) {
                 }
                 case "REWRITE_RESOURCE_PATH": {
                     removeAllQueries(requestContext, policy.getParameters());
+                    pathParamToQueryParamTransform(requestContext, policy.getParameters());
```
Comment (VirajSalaka) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java:94]: Do we need to restrict this to websocket APIs?

File: adapter/internal/oasparser/model/async_api.go [hunk 2]
```diff
@@ -136,14 +137,41 @@ func (asyncAPI AsyncAPI) getResources() []*Resource {
...
 		resource := unmarshalSwaggerResources(channel, methodsArray, channelItem.VendorExtensions)
 		resources = append(resources, &resource)
 	}
 
 	return SortResources(resources)
 }
 
+func populatePoliciesFromVendorExtensions(operation *Operation, vendorExtensions map[string]interface{}) {
+	if uriMapping, found := vendorExtensions[constants.XUriMapping]; found {
+		// TODO: (VirajSalaka) preprocess to process path param query param conversion.
```
Comment (VirajSalaka) [inline adapter/internal/oasparser/model/async_api.go:154]: remove unnecessary todo statement

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java [hunk 3]
```diff
@@ -143,6 +150,34 @@ private void removeAllQueries(RequestContext requestContext, Map<String, String>
         }
     }
 
+    private void pathParamToQueryParamTransform(RequestContext requestContext, Map<String, String> policyAttrib) {
+        if (policyAttrib.containsKey("x-uri-mapping")) {
```
Comment (VirajSalaka) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java:154]: introduce constant

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java (hunk 1 above)
Comment (VirajSalaka) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/interceptor/MediationPolicyFilter.java:94]: no need to do this as there is an if condition is inside the method

File: integration/mock-backend-server/src/main/java/org/wso2/choreo/connect/mockbackend/async/MockAsyncServer.java [hunk 4]
```diff
@@ -86,13 +83,14 @@ public void initChannel(SocketChannel ch) {
...
             pipeline.addLast(new WebSocketServerCompressionHandler());
-
-            // handle websocket handshake and the control frames (Close, Ping, Pong)
-            pipeline.addLast(new WebSocketServerProtocolHandler(WEBSOCKET_PATH, null, true));
-
-            // handle custom websocket implementation
-            pipeline.addLast(new WsServerFrameHandler());
-            log.info("Initialized pipeline to support WS at path {}", WEBSOCKET_PATH);
+            pipeline.addLast(new WsHttpRequestHandler());
+//
```
Comment (VirajSalaka) [inline integration/mock-backend-server/src/main/java/org/wso2/choreo/connect/mockbackend/async/MockAsyncServer.java:87]: Remove unnecessary code.

File: integration/mock-backend-server/src/main/java/org/wso2/choreo/connect/mockbackend/async/websocket/WsHttpRequestHandler.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+
+        if (msg instanceof HttpRequest) {
+
+            HttpRequest httpRequest = (HttpRequest) msg;
+            log.info("Http Request Received");
+
+            HttpHeaders headers = httpRequest.headers();
+            System.out.println("Connection : " +headers.get("Connection"));
```
Comment (VirajSalaka) [inline integration/mock-backend-server/src/main/java/org/wso2/choreo/connect/mockbackend/async/websocket/WsHttpRequestHandler.java:54]: log.info

**Comment by VirajSalaka** [comment]
Check for default versions

**Review by suksw — APPROVED** [review]

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc [hunk 6]
```diff
@@ -64,13 +64,18 @@ bool MgwWebSocketRootContext::onConfigure(size_t config_size) {
   return true;
 }
 
+MgwWebSocketContext::~MgwWebSocketContext(){
+  LOG_TRACE(std::string("MgwContext destructed") + this->x_request_id_);
```
Comment (AmaliMatharaarachchi) [inline envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc:68]: Just a query; Don't we have to delete stream_handler_ here?

File: enforcer-parent/commons/src/main/java/org/wso2/choreo/connect/enforcer/commons/model/RequestContext.java [hunk 7]
```diff
@@ -544,10 +544,19 @@ public Builder pathTemplate(String requestPathTemplate) {
...
+            // request.
+            if (!formattedRawPath.startsWith(formattedBasePath)) {
+                // formatted basePath does not have a trailing slash. Hence the last slash appears right
+                // before the version.
+                String apiContext = formattedBasePath.substring(0, formattedBasePath.lastIndexOf("/"));
+                completeResourcePathTemplate = apiContext + formattedResourcePathTemplate;
+            }
             final ParameterResolver parameterResolver = new ParameterResolver
-                    (formattedBasePath + formattedResourcePathTemplate);
+                    (completeResourcePathTemplate);
```
Comment (AmaliMatharaarachchi) [inline enforcer-parent/commons/src/main/java/org/wso2/choreo/connect/enforcer/commons/model/RequestContext.java:559]: formatting improvement - make it one line

File: enforcer-parent/commons/src/test/java/org/wso2/choreo/connect/enforcer/commons/model/RequestContextTest.java [hunk 8]
```diff
@@ -101,6 +101,12 @@ public void testPathParameterGenerationWithPrefixedPath() {
                 "petId", "12");
     }
 
+    @Test
+    public void testPathParameterWithDefaultAPIRequests() {
+        testPathParamValues("/testapi/pet/12", "/testapi/v1", "/pet/{petId}",
+                "petId", "12");
+    }
+
```
Comment (AmaliMatharaarachchi) [inline enforcer-parent/commons/src/test/java/org/wso2/choreo/connect/enforcer/commons/model/RequestContextTest.java:109]:
```suggestion
```

**Review by AmaliMatharaarachchi — APPROVED** [review]

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc (hunk 6 above)
Comment (VirajSalaka) [reply envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc:68]: Nope 
calling reset on stream_handler deletes the object. (I verified it with logs). Deleting explicitly leads to issues and not necessary.

**Review by pubudu538 — APPROVED** [review]

**Pull request merged**